### PR TITLE
Add support for variables

### PR DESCRIPTION
**Pull request by ltouati**
This PR add support for basic replacement of variables in skaffold.

The change is based on [viper](https://github.com/spf13/viper) and allow the use of a yaml file called env.yaml located either:
* in the /etc/skaffold/ directory
* in the $HOME/.skaffold directory
* in the directory where the skaffold file is located

Files will be applied on top of each others allowing overide

then in a skaffold file one can use
```
apiVersion: skaffold/v1beta2
kind: Config
build:
  artifacts:
  - image: ${image.registry.url}/${image.prefix}
    docker:
```
and given the following env.yaml file
```
image:
  registry:
    url: gcr.io/xxxx
  prefix:
    yyyy
```

will produce the following skaffold file
```
apiVersion: skaffold/v1beta2
kind: Config
build:
  artifacts:
  - image: gcr.io/xxxx/yyyy
    docker:
```

### CODE REVIEW

File: pkg/skaffold/util/util.go [hunk 1]
```diff
@@ -154,6 +154,68 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	targetDirectory := ""
+	if baseName != "skaffold.yaml" && baseName != "skaffold.yml" {
+		targetDirectory = pathToSkaffoldFile
+	} else {
+		targetDirectory = directory
+	}
+	logrus.Debugf("Will load environment file from :%s", targetDirectory)
+	ret := viper.New()
+	ret.SetConfigType("yaml")
+	ret.SetConfigName("env")
```
Comment (WoLfulus) [inline pkg/skaffold/util/util.go]: I personally want something that changes this to something else, like `skaffold --env prod.yaml`

File: pkg/skaffold/util/util.go [hunk 2]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	targetDirectory := ""
+	re := regexp.MustCompile("skaffold.*ya?ml")
+	if re.FindString(baseName) == "" {
+		targetDirectory = pathToSkaffoldFile
+	} else {
+		targetDirectory = directory
+	}
+	//
+	logrus.Debugf("Will load environment file from :%s", targetDirectory)
+	ret := viper.New()
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: can we call this `cfg` or something a bit more descriptive?

File: pkg/skaffold/util/util.go [hunk 3]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+		configurationValue := configurationSource.GetString(currentMatch)
+		logrus.Debugf("Configuration file replace of %s by %s", currentMatch, configurationValue)
+		return []byte(configurationValue)
+	})
+	logrus.Debugf("Replaced configuration file:%s", string(replacedBuffer))
+	return replacedBuffer
+}
+
+// Inject  environment variables into skaffold configuration
+func InjectEnvironnmentVariables(pathToSkaffoldFile string, skaffoldConfiguration []byte, profiles []string) ([]byte, error) {
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: `pathToSkaffoldFile` -> `configFilepath` or something similar

File: pkg/skaffold/util/util.go (hunk 1 above)
Comment (nkubala) [reply pkg/skaffold/util/util.go]: +1, this should be configurable through a CLI flag

File: pkg/skaffold/util/util.go [hunk 4]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+func FileExists(filapath string) bool {
+	if _, err := os.Stat(filapath); !os.IsNotExist(err) {
+		if err == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// Reads an environment file  that contains a set of environment variables
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: nit: extra space between `file` and `that`

Comment (nkubala) [reply pkg/skaffold/util/util.go]: also, could you expand your comment a bit to detail what is returned by this function

File: pkg/skaffold/util/util.go [hunk 5]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+// Inject  environment variables into skaffold configuration
+func InjectEnvironnmentVariables(pathToSkaffoldFile string, skaffoldConfiguration []byte, profiles []string) ([]byte, error) {
+	configurationSource, err := ReadEnvironmentConfigProperties(pathToSkaffoldFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "read environment config")
+	}
+	configurationSource.SetDefault("profile", strings.Join(profiles, ","))
+	return ReplaceEnvironmentVariables(configurationSource, skaffoldConfiguration), nil
+}
+func CopyStringMap(m map[string]string) map[string]string {
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: is this used?

File: pkg/skaffold/util/util.go [hunk 6]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	} else {
+		targetDirectory = directory
+	}
+	//
+	logrus.Debugf("Will load environment file from :%s", targetDirectory)
+	ret := viper.New()
+	ret.SetConfigType("yaml")
+	ret.SetConfigName("env")
+	ret.AddConfigPath("/etc/skaffold/")
+	ret.AddConfigPath("$HOME/.skaffold")
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: can you make these constants?

```golang
const defaultPropsPaths = []string{"/etc/skaffold/", "$HOME/.skaffold"}

for _, path := range defaultPropsPaths {
  cfg.AddConfigPath(path)
}
```

File: pkg/skaffold/util/util.go [hunk 7]
```diff
@@ -154,6 +154,76 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	return replacedBuffer
+}
+
+// Inject  environment variables into skaffold configuration
+func InjectEnvironnmentVariables(pathToSkaffoldFile string, skaffoldConfiguration []byte, profiles []string) ([]byte, error) {
+	configurationSource, err := ReadEnvironmentConfigProperties(pathToSkaffoldFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "read environment config")
+	}
+	configurationSource.SetDefault("profile", strings.Join(profiles, ","))
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: why do we need to set a value for `profile` here? shouldn't this get parsed normally through the replacer?

also, should it not be `profiles`?

Comment (ltouati) [reply pkg/skaffold/util/util.go]: yes makes sense, will remove it

File: pkg/skaffold/util/util.go [hunk 8]
```diff
@@ -154,6 +154,66 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	return false
+}
+
+var defaultPropsPaths = []string{"/etc/skaffold/", "$HOME/.skaffold"}
+
+// Reads an environment file that contains a set of environment variables
+// The file is expected to be in YAML. Once read, the function will
+// return an object that contains all the properties defined in the file
+func ReadEnvironmentConfigProperties(pathToEnvFile string) (*viper.Viper, error) {
+
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: nit: remove newline

File: pkg/skaffold/util/util.go [hunk 9]
```diff
@@ -154,6 +154,66 @@ func ReadConfiguration(filename string) ([]byte, error) {
...
+	logrus.Debugf("Will load environment file from :%s", pathToEnvFile)
+	configuration := viper.New()
+	for _, path := range defaultPropsPaths {
+		configuration.AddConfigPath(path)
```
Comment (nkubala) [inline pkg/skaffold/util/util.go]: could you explain how this works? it looks like you're adding default paths that the config files *might* live in, but then below you're explicitly calling `configuration.ReadConfig()` on a file descriptor created from the provided `pathToEnvFile`. are these other config files that might also get read if they exist?

**Comment by WoLfulus** [comment]
> Can you explain a little bit more about the actual missing features, individual parametrizations that are causing issues in your case?

I have to put my thoughts into this. Is there an open issue? Should we open one? I need compile all the points I felt skaffold was not helping before really commenting into this, but in resume I had to make a [full new cli project](https://github.com/reddotcloud/helmet) in which I generate skaffold file from my yaml and I pass it via pipe through `--filename -` to make skaffold work for my setup (even though I focused on helm in this case, I think skaffold could have helped more)

**Comment by nkubala** [comment]
@ltouati it looks like this PR has gone inactive and has several merge conflicts. I'm going to close this for now since I'm not sure we want to move in this direction - really sorry to let this one sit for long.

@WoLfulus I'd love to hear a little more about where skaffold isn't satisfying your workflows. would you mind opening some issues for the bigger concerns you have and we can move the discussion there? we're trying hard to improve the UX for many different use cases so I'm sure we can find ways to help you (and others) out :smiley:

**Pull request closed without merging**